universe/src/server.rs

```diff
@@ -0,0 +1,210 @@
+    }
+
+    #[instrument(skip(self))]
+    async fn list_keyspaces(
+        &self,
+        request: Request<ListKeyspacesRequest>,
+    ) -> Result<Response<ListKeyspacesResponse>, Status> {
+        debug!("Got a list_keyspaces request: {:?}", request);
+
+        let keyspaces = self
+            .storage
+            .list_keyspaces()
+            .await
+            .map_err(|e| Status::internal(format!("Failed to list keyspaces: {}", e)))?;
+
+        let response = ListKeyspacesResponse { keyspaces };
+        Ok(Response::new(response))
+    }
+}
+
+/// Runs the Universe Manager, listening on the provided address.
+///
+/// # Arguments
+/// - `addr`: The address for the Universe server to listen on.
+///
+/// # Errors
+/// This function will return an error if the provided address is invalid or if there is an issue
+/// starting the Universe server.
+pub async fn run_universe_server<S: Storage>(
+    addr: impl AsRef<str> + std::net::ToSocketAddrs,
+    storage: S,
+) -> Result<(), Box<dyn std::error::Error>> {
+    let addr = addr.to_socket_addrs()?.next().ok_or("Invalid address")?;
+
+    info!("UniverseServer listening on {}", addr);
+
+    let universe_server = UniverseServer::new(Arc::new(storage));
+
+    tonic::transport::Server::builder()
+        .add_service(proto::universe::universe_server::UniverseServer::new(
+            universe_server,
+        ))
+        .serve(addr)
+        .await?;
+
+    Ok(())
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use crate::storage::cassandra::Cassandra;
+    use proto::universe::{KeyRange, KeyRangeRequest};
+    use std::time::Duration;
+    use tokio::time::sleep;
+
+    #[tokio::test]
+    async fn test_run_universe_server() {
+        // Use a random available port
+        let addr = "127.0.0.1:0";
+
+        // Create a Cassandra CQL session
+        let storage = Cassandra::new("127.0.0.1:9042".to_string()).await;
+
+        // Start the server in a separate task
+        let server_task = tokio::spawn(async move {
+            run_universe_server(addr, storage).await.unwrap();
+        });
+
+        // Wait for 1 second
+        sleep(Duration::from_secs(1)).await;
+
+        // Stop the server by aborting the task
+        server_task.abort();
+
+        // Wait for the server to shut down
+        let _ = server_task.await;
+    }
+
+    #[tokio::test]
+    async fn test_create_and_list_keyspace_handlers() {
+        let addr = "127.0.0.1:50056";
+
+        // Create a Cassandra CQL session
+        let storage = Cassandra::new("127.0.0.1:9042".to_string()).await;
+
+        // Start the server in a separate task
+        let server_task = tokio::spawn(async move {
+            run_universe_server(addr, storage).await.unwrap();
+        });
+
+        // Wait a bit for the server to start
+        sleep(Duration::from_millis(100)).await;
+
+        // Create a Universe client
+        let mut client = UniverseClient::connect(format!("http://{}", addr))
```

Comment:
It's great to test using a proto client, but let's do that in a separate integration_tests file. Unit tests can just use the functions directly on the server